File: app/models/exam/Grade.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package models.exam;

import com.fasterxml.jackson.annotation.JsonBackReference;
import io.ebean.Model;
import io.ebean.annotation.EnumValue;
import jakarta.persistence.*;
import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;

@Entity
public class Grade extends Model {

    public enum Type {
        @EnumValue("1")
        GRADED,
        @EnumValue("2")
        NOT_GRADED,
        @EnumValue("3")
        POINT_GRADED,
    }

    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE)
    private Integer id;

    @Column
    private String name;

    @Column
    private Boolean marksRejection;

    @ManyToOne
    @JsonBackReference
    private GradeScale gradeScale;

    public Integer getId() {
        return id;
    }

    public void setId(Integer id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Boolean getMarksRejection() {
        return marksRejection;
    }

    public void setMarksRejection(Boolean marksRejection) {
        this.marksRejection = marksRejection;
    }

    public GradeScale getGradeScale() {
        return gradeScale;
    }

    public void setGradeScale(GradeScale gradeScale) {
        this.gradeScale = gradeScale;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        if (!(other instanceof Grade otherGrade)) {
            return false;
        }
        return new EqualsBuilder().append(id, otherGrade.id).build();
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder().append(id).build();
    }
}

File: app/models/exam/Grade.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package models.exam

import com.fasterxml.jackson.annotation.JsonBackReference
import io.ebean.Model
import jakarta.persistence.*

import scala.compiletime.uninitialized

@Entity
class Grade extends Model:
  @Id
  @GeneratedValue(strategy = GenerationType.SEQUENCE)
  var id: Integer = uninitialized

  @ManyToOne
  @JsonBackReference
  var gradeScale: GradeScale = uninitialized

  var name: String                      = uninitialized
  var marksRejection: java.lang.Boolean = uninitialized

  override def equals(o: Any): Boolean = o match
    case g: Grade => this.id == g.id
    case _        => false

  override def hashCode: Int = if id != null then id.hashCode else 0

object Grade:
  type Type = GradeType
  val Type: Class[GradeType] = classOf[GradeType]
